File: or-sdk/libs/EMInterface/src/utilities.jl
"""
    merge_exp_name!(search_args, experimentName)

Ensure `search_args` is a `Dict{String, Any}` and add `experimentName`.
"""
function merge_exp_name!(search_args, experimentName)
    search_args = convert(Dict{String, Any}, search_args)
    merge!(search_args, Dict("experimentName" => experimentName))
    return search_args
end

"""
    check_valid_search(search_args::Dict, query::String)

Error if any keys of `search_args` are not allowed to be used in `query`. If a full
introspection of the GraqphQL server has been completed, the arguments are compared
to the schema. If not, they are compared to hardcoded lists in this function.
"""
function check_valid_search(search_args::Dict, query::String)
    if get_client().introspection_complete
        @debug "Using introspected GQL server schema to check allowed search fields."
        allowed_keys = keys(get_client().query_to_args_map[query])
    else
        @debug "GQL schema not available, using hardcoded lists of allowed search fields."
        if query == "getSimulation"
            allowed_keys = ("experimentName", "primaryKey", "simName", "objectId", "p", "timestamp")
        elseif query == "getSimulationAgent"
            allowed_keys = ("experimentName", "primaryKey", "simName", "objectId")
        elseif query == "getConfigurations"
            allowed_keys = ("experimentName", "studyName", "experimentBase", "configParent",
                            "createdBy")
        else
            throw(ExperimentManagerException("No available search fields for query \"$query\".\nIf not expected, try running full_introspection() to get schema"))
        end
    end

    map(collect(keys(search_args))) do key
        if !(key in allowed_keys)
            throw(ExperimentManagerException("Cannot search simulation results on field \"$key\""))
        end
    end
end

"""
    macro async_display_errors expr

Wraps `expr` in a try catch and runs it aysnchronously. If an error is thrown,
this is displayed as a warning.

# Examples
```julia
julia> @async_display_errors foo(bar)
```
"""
macro async_display_errors(expr)
    caught_expr = quote
        @async try
            $(esc(expr))
        catch err
            @error "Async function failed: $(typeof(err))" exception=(err, catch_backtrace())
        end
    end
    return caught_expr
end

######################
# sec_into_day_calcs #
######################
# TODO: Since these calculations are now centralised, remove duplication in other services.

"""
    round_to_nearest(num, base)

Round to nearerst `base`.

# Examples

```julia
julia> round_to_nearest(1,10)
0
julia> round_to_nearest(11,10)
10
```
"""
round_to_nearest(num, base) = base * round(Int, num / base)

"""
    floor_to_nearest(num, base)

Floor to nearerst `base` to ensure numbers align for different bases.

# Examples

```julia
julia> floor_to_nearest(1,10)
0
julia> floor_to_nearest(6,10)
0
julia> floor_to_nearest(11,10)
10
julia> floor_to_nearest(11,10)
10
```
"""
floor_to_nearest(num, base) = base * floor(Int, num / base)


"""
    sec_into_day_from_timestamp(timestamp::Integer, frequency::Integer)
    sec_into_day_from_timestamp(timestamp::Integer, frequency)
    sec_into_day_from_timestamp(timestamp::Float64, frequency)

From a unix timestamp, get the seconds into day floored to the frequency specified.
"""
sec_into_day_from_timestamp(timestamp::Integer, frequency::Integer) = floor_to_nearest(timestamp % 86400, frequency)
sec_into_day_from_timestamp(timestamp::Integer, frequency) = floor_to_nearest(timestamp % 86400, floor(Int, frequency))
sec_into_day_from_timestamp(timestamp::Float64, frequency) = sec_into_day_from_timestamp(floor(Int, timestamp), floor(Int, frequency))

File: or-sdk/libs/em_interface_python/em_interface/utilities.py
import logging

from em_interface.connect import get_client
from em_interface.exceptions import ExperimentManagerException

logger = logging.getLogger(__name__)

def merge_exp_name(search_args, experiment_name):
    """
    Adds experiment_name to the search arguments.

    Args
    ----
    `search_args: Dict{String, Any}`

    `experiment_name: String`

    Returns
    -------
    `search_args: Dict{String, Any}`

    """
    search_args["experimentName"] = experiment_name
    return search_args


def check_valid_search(search_args, query):
    """
    Checks that the provided search arguments are available for the provided query.

    Args
    ----
    `search_args: Dict{String, Any}`

    `query: String`

    Returns
    -------
    None.

    """
    client = get_client()
    if client.introspection_complete:
        allowed_keys = client.query_to_args_map[query].keys()
    else:
        logger.debug("GQL schema not available, using hardcoded lists of allowed search fields.")
        if query == "getSimulation":
            allowed_keys = ("experimentName", "primaryKey", "simName", "objectId", "p", "timestamp")
        elif query == "getSimulationAgent":
            allowed_keys = ("experimentName", "primaryKey", "simName", "objectId")
        elif query == "getConfigurations":
            allowed_keys = ("experimentName", "studyName", "experimentBase", "configParent", "createdBy")
        else:
            raise ExperimentManagerException(f"No available search fields for query {query}.\n If not expected, try running full_introspection() to get schema ")

    for key in search_args.keys():
        if key not in allowed_keys:
            raise ExperimentManagerException(f"Cannot search {query} on field {key}")


def build_query_args(search_args:dict):
    """
    Add `search_args` dictionary to a new dictionary, as the value for the key `search`.

    The majority of the Experiment Manager queries have this search argument, so this utility function avoids repeated\
        code.

    Args
    ----
    `search_args:dict`

    Returns
    -------
    `Bool` or `Dict`

    """
    query_args = search_args if bool(search_args) else {"search": search_args}
    return query_args